PcesReadWriteTests.corruptedEventTest failing on seed -5596113931140436367L

https://github.com/hashgraph/hedera-services/blob/70c450e1692f67a7aae26a75b9cd3402af7d209a/platform-sdk/swirlds-unit-tests/core/swirlds-platform-test/src/test/java/com/swirlds/platform/test/event/preconsensus/PcesReadWriteTests.java#L325-L327

when run with 
```
 final Random random = RandomUtils.getRandomPrintSeed(-5596113931140436367L);
```
Fails both the GENERATION_THRESHOLD and BIRTH_ROUND_THRESHOLD

**Background** 

We run most of our tests based on randomized seeds to discover corner cases.  This seed has discovered a corner case.  Running the unit test without the seed specified passes most runs. 

The error message: 
```

org.opentest4j.AssertionFailedError: Unexpected exception type thrown, expected: <java.io.IOException> but was: <java.util.NoSuchElementException>
	at org.junit.jupiter.api.AssertionFailureBuilder.build(AssertionFailureBuilder.java:151)
	at org.junit.jupiter.api.AssertThrows.assertThrows(AssertThrows.java:67)
	at org.junit.jupiter.api.AssertThrows.assertThrows(AssertThrows.java:35)
	at org.junit.jupiter.api.Assertions.assertThrows(Assertions.java:3115)
	at com.swirlds.platform.test.event.preconsensus.PcesReadWriteTests.corruptedEventsTest(PcesReadWriteTests.java:383)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.accept(ForEachOps.java:184)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:197)
	at java.base/java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:179)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:197)
	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.accept(ForEachOps.java:184)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:197)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:197)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:197)
	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.accept(ForEachOps.java:184)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:197)
	at java.base/java.util.Spliterators$ArraySpliterator.forEachRemaining(Spliterators.java:1024)
	at java.base/java.util.stream.ReferencePipeline$Head.forEach(ReferencePipeline.java:762)
	at java.base/java.util.stream.ReferencePipeline$7$1.accept(ReferencePipeline.java:276)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:197)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:197)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:197)
	at java.base/java.util.Spliterators$ArraySpliterator.forEachRemaining(Spliterators.java:1024)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:509)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:499)
	at java.base/java.util.stream.ForEachOps$ForEachOp.evaluateSequential(ForEachOps.java:151)
	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(ForEachOps.java:174)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.base/java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:596)
	at java.base/java.util.stream.ReferencePipeline$7$1.accept(ReferencePipeline.java:276)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:197)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:197)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:197)
	at java.base/java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1708)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:509)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:499)
	at java.base/java.util.stream.ForEachOps$ForEachOp.evaluateSequential(ForEachOps.java:151)
	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(ForEachOps.java:174)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.base/java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:596)
	at java.base/java.util.stream.ReferencePipeline$7$1.accept(ReferencePipeline.java:276)
	at java.base/java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1708)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:509)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:499)
	at java.base/java.util.stream.ForEachOps$ForEachOp.evaluateSequential(ForEachOps.java:151)
	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(ForEachOps.java:174)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.base/java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:596)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1596)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1596)
Caused by: java.util.NoSuchElementException: no files remain in this iterator
	at com.swirlds.platform.event.preconsensus.PcesFileIterator.next(PcesFileIterator.java:142)
	at org.junit.jupiter.api.AssertThrows.assertThrows(AssertThrows.java:53)
```

Need to trace through the code to figure out why there are no remaining files in the iterator.  Did the unit test fail to create an invalid event?  Was it created at the boundary of a file and not read?   

Comparison to other seed runs will be important to understand how the test should be failing. 


**Acceptance Criteria**

This issue is complete when either: 
1. A PR is created to fix the problem. 
2. The cause of the failure is discovered and documented in a new issue which will be closed by a PR fixing the issue.

--- a/platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/event/preconsensus/PcesFileIterator.java
+++ b/platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/event/preconsensus/PcesFileIterator.java
@@ -18,8 +18,6 @@ package com.swirlds.platform.event.preconsensus;
 
 import com.hedera.hapi.platform.event.GossipEvent;
 import com.swirlds.common.io.IOIterator;
-import com.swirlds.common.io.extendable.ExtendableInputStream;
-import com.swirlds.common.io.extendable.extensions.CountingStreamExtension;
 import com.swirlds.common.io.streams.SerializableDataInputStream;
 import com.swirlds.platform.event.AncientMode;
 import com.swirlds.platform.event.EventSerializationUtils;
@@ -41,7 +39,6 @@ public class PcesFileIterator implements IOIterator<PlatformEvent> {
     private final AncientMode fileType;
     private final SerializableDataInputStream stream;
     private boolean hasPartialEvent = false;
-    private final CountingStreamExtension counter;
     private PlatformEvent next;
     private boolean streamClosed = false;
     private PcesFileVersion fileVersion;
@@ -60,11 +57,9 @@ public class PcesFileIterator implements IOIterator<PlatformEvent> {
 
         this.lowerBound = lowerBound;
         this.fileType = Objects.requireNonNull(fileType);
-        counter = new CountingStreamExtension();
-        stream = new SerializableDataInputStream(new ExtendableInputStream(
+        stream = new SerializableDataInputStream(
                 new BufferedInputStream(
-                        new FileInputStream(fileDescriptor.getPath().toFile())),
-                counter));
+                        new FileInputStream(fileDescriptor.getPath().toFile())));
 
         try {
             final int fileVersionNumber = stream.readInt();
@@ -74,8 +69,7 @@ public class PcesFileIterator implements IOIterator<PlatformEvent> {
             }
         } catch (final EOFException e) {
             // Empty file. Possible if the node crashed right after it created this file.
-            stream.close();
-            streamClosed = true;
+            closeFile();
         }
     }
 
@@ -84,35 +78,38 @@ public class PcesFileIterator implements IOIterator<PlatformEvent> {
      */
     private void findNext() throws IOException {
         while (next == null && !streamClosed) {
-
-            final long initialCount = counter.getCount();
+            if (stream.available() == 0) {
+                closeFile();
+                return;
+            }
 
             try {
                 final PlatformEvent candidate =
                         switch (fileVersion) {
                             case ORIGINAL -> EventSerializationUtils.deserializePlatformEvent(stream, true);
-                            case PROTOBUF_EVENTS -> {
-                                final GossipEvent gossipEvent = stream.readPbjRecord(GossipEvent.PROTOBUF);
-                                try {
-                                    yield new PlatformEvent(gossipEvent);
-                                } catch (final NullPointerException e) {
-                                    throw new IOException("GossipEvent read from the file is malformed", e);
-                                }
-                            }
+                            case PROTOBUF_EVENTS -> new PlatformEvent(stream.readPbjRecord(GossipEvent.PROTOBUF));
                         };
                 if (candidate.getAncientIndicator(fileType) >= lowerBound) {
                     next = candidate;
                 }
-            } catch (final EOFException e) {
-                if (counter.getCount() > initialCount) {
-                    // We started parsing an event but couldn't find enough bytes to finish it.
-                    // This is possible (if not likely) when a node is shut down abruptly.
-                    hasPartialEvent = true;
-                }
-                stream.close();
-                streamClosed = true;
+            } catch (final IOException e) {
+                // We started parsing an event but couldn't find enough bytes to finish it.
+                // This is possible (if not likely) when a node is shut down abruptly.
+                hasPartialEvent = true;
+                closeFile();
+                throw e;
+            } catch (final NullPointerException e) {
+                // The PlatformEvent constructor can throw this if the event is malformed.
+                hasPartialEvent = true;
+                closeFile();
+                throw new IOException("GossipEvent read from the file is malformed", e);
             }
         }
+    }
+
+    private void closeFile() throws IOException {
+        stream.close();
+        streamClosed = true;
     }
 
     /**

--- a/platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/event/preconsensus/PcesFileIterator.java
+++ b/platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/event/preconsensus/PcesFileIterator.java
@@ -1,5 +1,5 @@
 /*
- * Copyright (C) 2016-2024 Hedera Hashgraph, LLC
+ * Copyright (C) 2016-2025 Hedera Hashgraph, LLC
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
@@ -57,9 +57,8 @@ public class PcesFileIterator implements IOIterator<PlatformEvent> {
 
         this.lowerBound = lowerBound;
         this.fileType = Objects.requireNonNull(fileType);
-        stream = new SerializableDataInputStream(
-                new BufferedInputStream(
-                        new FileInputStream(fileDescriptor.getPath().toFile())));
+        stream = new SerializableDataInputStream(new BufferedInputStream(
+                new FileInputStream(fileDescriptor.getPath().toFile())));
 
         try {
             final int fileVersionNumber = stream.readInt();

--- a/platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/event/preconsensus/PcesMultiFileIterator.java
+++ b/platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/event/preconsensus/PcesMultiFileIterator.java
@@ -1,5 +1,5 @@
 /*
- * Copyright (C) 2016-2024 Hedera Hashgraph, LLC
+ * Copyright (C) 2016-2025 Hedera Hashgraph, LLC
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
@@ -76,7 +76,7 @@ public class PcesMultiFileIterator implements IOIterator<PlatformEvent> {
                 // ignore the exception and move on to the next file if there is one
             }
 
-            if(hasNextEvent){
+            if (hasNextEvent) {
                 next = currentIterator.next();
                 return;
             }

--- a/platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/event/preconsensus/PcesMultiFileIterator.java
+++ b/platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/event/preconsensus/PcesMultiFileIterator.java
@@ -59,20 +59,37 @@ public class PcesMultiFileIterator implements IOIterator<PlatformEvent> {
      * Find the next event that should be returned.
      */
     private void findNext() throws IOException {
-        while (next == null) {
-            if (currentIterator == null || !currentIterator.hasNext()) {
-                if (currentIterator != null && currentIterator.hasPartialEvent()) {
-                    truncatedFileCount++;
-                }
-
-                if (!fileIterator.hasNext()) {
-                    break;
-                }
-
+        if (currentIterator == null) { // on first call
+            if (fileIterator.hasNext()) {
                 currentIterator = new PcesFileIterator(fileIterator.next(), lowerBound, fileType);
             } else {
-                next = currentIterator.next();
+                return;
             }
+        }
+
+        while (next == null) {
+
+            boolean hasNextEvent = false;
+            try {
+                hasNextEvent = currentIterator.hasNext();
+            } catch (final IOException ignored) {
+                // ignore the exception and move on to the next file if there is one
+            }
+
+            if(hasNextEvent){
+                next = currentIterator.next();
+                return;
+            }
+
+            if (currentIterator.hasPartialEvent()) {
+                truncatedFileCount++;
+            }
+
+            if (!fileIterator.hasNext()) {
+                return;
+            }
+
+            currentIterator = new PcesFileIterator(fileIterator.next(), lowerBound, fileType);
         }
     }
 

--- a/platform-sdk/swirlds-platform-core/src/testFixtures/java/com/swirlds/platform/test/fixtures/event/PcesWriterTestUtils.java
+++ b/platform-sdk/swirlds-platform-core/src/testFixtures/java/com/swirlds/platform/test/fixtures/event/PcesWriterTestUtils.java
@@ -1,5 +1,5 @@
 /*
- * Copyright (C) 2024 Hedera Hashgraph, LLC
+ * Copyright (C) 2024-2025 Hedera Hashgraph, LLC
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
@@ -22,7 +22,6 @@ import static com.swirlds.common.utility.CompareTo.isGreaterThanOrEqualTo;
 import static com.swirlds.platform.system.transaction.TransactionWrapperUtils.createAppPayloadWrapper;
 import static org.junit.jupiter.api.Assertions.assertEquals;
 import static org.junit.jupiter.api.Assertions.assertFalse;
-import static org.junit.jupiter.api.Assertions.assertThrows;
 import static org.junit.jupiter.api.Assertions.assertTrue;
 
 import com.swirlds.common.context.PlatformContext;
@@ -175,7 +174,7 @@ public class PcesWriterTestUtils {
                     assertTrue(event.getAncientIndicator(ancientMode) >= file.getLowerBound());
                     assertTrue(event.getAncientIndicator(ancientMode) <= file.getUpperBound());
                 }
-            }catch (final IOException ignored){
+            } catch (final IOException ignored) {
                 // hasNext() can throw an IOException if the file is truncated, in this case there is nothing to do
             }
         }

--- a/platform-sdk/swirlds-platform-core/src/testFixtures/java/com/swirlds/platform/test/fixtures/event/PcesWriterTestUtils.java
+++ b/platform-sdk/swirlds-platform-core/src/testFixtures/java/com/swirlds/platform/test/fixtures/event/PcesWriterTestUtils.java
@@ -22,6 +22,7 @@ import static com.swirlds.common.utility.CompareTo.isGreaterThanOrEqualTo;
 import static com.swirlds.platform.system.transaction.TransactionWrapperUtils.createAppPayloadWrapper;
 import static org.junit.jupiter.api.Assertions.assertEquals;
 import static org.junit.jupiter.api.Assertions.assertFalse;
+import static org.junit.jupiter.api.Assertions.assertThrows;
 import static org.junit.jupiter.api.Assertions.assertTrue;
 
 import com.swirlds.common.context.PlatformContext;
@@ -125,8 +126,7 @@ public class PcesWriterTestUtils {
             assertTrue(eventsIterator.hasNext());
             assertEquals(event, eventsIterator.next());
         }
-
-        assertFalse(eventsIterator.hasNext());
+        assertFalse(eventsIterator.hasNext(), "There should be no more events");
         assertEquals(truncatedFileCount, eventsIterator.getTruncatedFileCount());
 
         // Make sure things look good when iterating starting in the middle of the stream that was written
@@ -169,11 +169,14 @@ public class PcesWriterTestUtils {
             assertTrue(file.getUpperBound() >= previousMaximum);
             previousMaximum = file.getUpperBound();
 
-            final IOIterator<PlatformEvent> fileEvents = file.iterator(0);
-            while (fileEvents.hasNext()) {
-                final PlatformEvent event = fileEvents.next();
-                assertTrue(event.getAncientIndicator(ancientMode) >= file.getLowerBound());
-                assertTrue(event.getAncientIndicator(ancientMode) <= file.getUpperBound());
+            try (final IOIterator<PlatformEvent> fileEvents = file.iterator(0)) {
+                while (fileEvents.hasNext()) {
+                    final PlatformEvent event = fileEvents.next();
+                    assertTrue(event.getAncientIndicator(ancientMode) >= file.getLowerBound());
+                    assertTrue(event.getAncientIndicator(ancientMode) <= file.getUpperBound());
+                }
+            }catch (final IOException ignored){
+                // hasNext() can throw an IOException if the file is truncated, in this case there is nothing to do
             }
         }
     }

--- a/platform-sdk/swirlds-unit-tests/core/swirlds-platform-test/src/test/java/com/swirlds/platform/test/event/preconsensus/PcesReadWriteTests.java
+++ b/platform-sdk/swirlds-unit-tests/core/swirlds-platform-test/src/test/java/com/swirlds/platform/test/event/preconsensus/PcesReadWriteTests.java
@@ -1,5 +1,5 @@
 /*
- * Copyright (C) 2016-2024 Hedera Hashgraph, LLC
+ * Copyright (C) 2016-2025 Hedera Hashgraph, LLC
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
@@ -304,7 +304,9 @@ class PcesReadWriteTests {
         if (truncateOnBoundary) {
             iterator.forEachRemaining(deserializedEvents::add);
         } else {
-            assertThrows(IOException.class, () -> iterator.forEachRemaining(deserializedEvents::add),
+            assertThrows(
+                    IOException.class,
+                    () -> iterator.forEachRemaining(deserializedEvents::add),
                     "A partial event should have been detected and an IOException should have been thrown");
         }
 
@@ -315,7 +317,6 @@ class PcesReadWriteTests {
         for (int i = 0; i < deserializedEvents.size(); i++) {
             assertEquals(events.get(i), deserializedEvents.get(i));
         }
-
     }
 
     @ParameterizedTest

--- a/platform-sdk/swirlds-unit-tests/core/swirlds-platform-test/src/test/java/com/swirlds/platform/test/event/preconsensus/PcesReadWriteTests.java
+++ b/platform-sdk/swirlds-unit-tests/core/swirlds-platform-test/src/test/java/com/swirlds/platform/test/event/preconsensus/PcesReadWriteTests.java
@@ -49,7 +49,6 @@ import java.nio.file.Path;
 import java.time.Instant;
 import java.util.ArrayList;
 import java.util.HashMap;
-import java.util.Iterator;
 import java.util.List;
 import java.util.Map;
 import java.util.NoSuchElementException;
@@ -100,7 +99,7 @@ class PcesReadWriteTests {
         return Stream.of(Arguments.of(GENERATION_THRESHOLD), Arguments.of(BIRTH_ROUND_THRESHOLD));
     }
 
-    protected static Stream<Arguments> ancientAndWriterTypeArguments() {
+    protected static Stream<Arguments> ancientAndBoolanArguments() {
         return Stream.of(
                 Arguments.of(GENERATION_THRESHOLD, false),
                 Arguments.of(BIRTH_ROUND_THRESHOLD, false),
@@ -109,7 +108,7 @@ class PcesReadWriteTests {
     }
 
     @ParameterizedTest
-    @MethodSource("ancientAndWriterTypeArguments")
+    @MethodSource("ancientAndBoolanArguments")
     @DisplayName("Write Then Read Test")
     void writeThenReadTest(@NonNull final AncientMode ancientMode, final boolean useFileChannelWriter)
             throws IOException {
@@ -213,13 +212,7 @@ class PcesReadWriteTests {
         iterator.forEachRemaining(deserializedEvents::add);
 
         // We don't want any events with an ancient indicator less than the middle
-        final Iterator<PlatformEvent> it = events.iterator();
-        while (it.hasNext()) {
-            final PlatformEvent event = it.next();
-            if (event.getAncientIndicator(ancientMode) < middle) {
-                it.remove();
-            }
-        }
+        events.removeIf(event -> event.getAncientIndicator(ancientMode) < middle);
 
         assertEquals(events.size(), deserializedEvents.size());
         for (int i = 0; i < events.size(); i++) {
@@ -250,73 +243,79 @@ class PcesReadWriteTests {
     }
 
     @ParameterizedTest
-    @MethodSource("ancientModeArguments")
+    @MethodSource("ancientAndBoolanArguments")
     @DisplayName("Truncated Event Test")
-    void truncatedEventTest(@NonNull final AncientMode ancientMode) throws IOException {
-        for (final boolean truncateOnBoundary : List.of(true, false)) {
-            final Random random = RandomUtils.getRandomPrintSeed();
+    void truncatedEventTest(@NonNull final AncientMode ancientMode, final boolean truncateOnBoundary)
+            throws IOException {
+        final Random random = RandomUtils.getRandomPrintSeed();
 
-            final int numEvents = 100;
+        final int numEvents = 100;
 
-            final StandardGraphGenerator generator = new StandardGraphGenerator(
-                    ancientMode == GENERATION_THRESHOLD ? DEFAULT_PLATFORM_CONTEXT : BIRTH_ROUND_PLATFORM_CONTEXT,
-                    random.nextLong(),
-                    new StandardEventSource(),
-                    new StandardEventSource(),
-                    new StandardEventSource(),
-                    new StandardEventSource());
+        final StandardGraphGenerator generator = new StandardGraphGenerator(
+                ancientMode == GENERATION_THRESHOLD ? DEFAULT_PLATFORM_CONTEXT : BIRTH_ROUND_PLATFORM_CONTEXT,
+                random.nextLong(),
+                new StandardEventSource(),
+                new StandardEventSource(),
+                new StandardEventSource(),
+                new StandardEventSource());
 
-            final List<PlatformEvent> events = new ArrayList<>();
-            for (int i = 0; i < numEvents; i++) {
-                events.add(generator.generateEvent().getBaseEvent());
-            }
-
-            long upperBound = Long.MIN_VALUE;
-            for (final PlatformEvent event : events) {
-                upperBound = Math.max(upperBound, event.getAncientIndicator(ancientMode));
-            }
-
-            upperBound += random.nextInt(0, 10);
-
-            final PcesFile file = PcesFile.of(
-                    ancientMode,
-                    RandomUtils.randomInstant(random),
-                    random.nextInt(0, 100),
-                    0,
-                    upperBound,
-                    upperBound,
-                    testDirectory);
-
-            final Map<Integer /* event index */, Integer /* last byte position */> byteBoundaries = new HashMap<>();
-
-            final PcesMutableFile mutableFile = file.getMutableFile();
-            for (int i = 0; i < events.size(); i++) {
-                final PlatformEvent event = events.get(i);
-                mutableFile.writeEvent(event);
-                byteBoundaries.put(i, (int) mutableFile.fileSize());
-            }
-
-            mutableFile.close();
-
-            final int lastEventIndex =
-                    random.nextInt(0, events.size() - 2 /* make sure we always truncate at least one event */);
-
-            final int truncationPosition = byteBoundaries.get(lastEventIndex) + (truncateOnBoundary ? 0 : 1);
-
-            truncateFile(file.getPath(), truncationPosition);
-
-            final PcesFileIterator iterator = file.iterator(Long.MIN_VALUE);
-            final List<PlatformEvent> deserializedEvents = new ArrayList<>();
-            iterator.forEachRemaining(deserializedEvents::add);
-
-            assertEquals(truncateOnBoundary, !iterator.hasPartialEvent());
-
-            assertEquals(lastEventIndex + 1, deserializedEvents.size());
-
-            for (int i = 0; i < deserializedEvents.size(); i++) {
-                assertEquals(events.get(i), deserializedEvents.get(i));
-            }
+        final List<PlatformEvent> events = new ArrayList<>();
+        for (int i = 0; i < numEvents; i++) {
+            events.add(generator.generateEvent().getBaseEvent());
         }
+
+        long upperBound = Long.MIN_VALUE;
+        for (final PlatformEvent event : events) {
+            upperBound = Math.max(upperBound, event.getAncientIndicator(ancientMode));
+        }
+
+        upperBound += random.nextInt(0, 10);
+
+        final PcesFile file = PcesFile.of(
+                ancientMode,
+                RandomUtils.randomInstant(random),
+                random.nextInt(0, 100),
+                0,
+                upperBound,
+                upperBound,
+                testDirectory);
+
+        final Map<Integer /* event index */, Integer /* last byte position */> byteBoundaries = new HashMap<>();
+
+        final PcesMutableFile mutableFile = file.getMutableFile();
+        for (int i = 0; i < events.size(); i++) {
+            final PlatformEvent event = events.get(i);
+            mutableFile.writeEvent(event);
+            byteBoundaries.put(i, (int) mutableFile.fileSize());
+        }
+
+        mutableFile.close();
+
+        final int lastEventIndex =
+                random.nextInt(0, events.size() - 2 /* make sure we always truncate at least one event */);
+
+        final int truncationPosition = byteBoundaries.get(lastEventIndex) + (truncateOnBoundary ? 0 : 1);
+
+        truncateFile(file.getPath(), truncationPosition);
+
+        final PcesFileIterator iterator = file.iterator(Long.MIN_VALUE);
+        final List<PlatformEvent> deserializedEvents = new ArrayList<>();
+
+        if (truncateOnBoundary) {
+            iterator.forEachRemaining(deserializedEvents::add);
+        } else {
+            assertThrows(IOException.class, () -> iterator.forEachRemaining(deserializedEvents::add),
+                    "A partial event should have been detected and an IOException should have been thrown");
+        }
+
+        assertEquals(truncateOnBoundary, !iterator.hasPartialEvent());
+
+        assertEquals(lastEventIndex + 1, deserializedEvents.size());
+
+        for (int i = 0; i < deserializedEvents.size(); i++) {
+            assertEquals(events.get(i), deserializedEvents.get(i));
+        }
+
     }
 
     @ParameterizedTest
@@ -593,8 +592,9 @@ class PcesReadWriteTests {
         assertTrue(path.toFile().createNewFile());
         assertTrue(Files.exists(path));
 
-        final PcesFileIterator iterator = file.iterator(Long.MIN_VALUE);
-        assertFalse(iterator.hasNext());
-        assertThrows(NoSuchElementException.class, iterator::next);
+        try (final PcesFileIterator iterator = file.iterator(Long.MIN_VALUE)) {
+            assertFalse(iterator.hasNext());
+            assertThrows(NoSuchElementException.class, iterator::next);
+        }
     }
 }

--- a/platform-sdk/swirlds-unit-tests/core/swirlds-platform-test/src/test/java/com/swirlds/platform/test/event/preconsensus/PcesWriterTests.java
+++ b/platform-sdk/swirlds-unit-tests/core/swirlds-platform-test/src/test/java/com/swirlds/platform/test/event/preconsensus/PcesWriterTests.java
@@ -81,6 +81,7 @@ import org.junit.jupiter.api.DisplayName;
 import org.junit.jupiter.api.io.TempDir;
 import org.junit.jupiter.params.ParameterizedTest;
 import org.junit.jupiter.params.provider.Arguments;
+import org.junit.jupiter.params.provider.CsvSource;
 import org.junit.jupiter.params.provider.MethodSource;
 
 @DisplayName("PcesWriter Tests")
@@ -399,125 +400,129 @@ class PcesWriterTests {
     }
 
     @ParameterizedTest
-    @MethodSource("buildArguments")
+    @CsvSource({
+            "GENERATION_THRESHOLD, true",
+            "GENERATION_THRESHOLD, false",
+            "BIRTH_ROUND_THRESHOLD, true",
+            "BIRTH_ROUND_THRESHOLD, false"
+    })
     @DisplayName("Discontinuity Test")
-    void discontinuityTest(@NonNull final AncientMode ancientMode) throws IOException {
-        for (final boolean truncateLastFile : List.of(true, false)) {
-            beforeEach();
+    void discontinuityTest(@NonNull final AncientMode ancientMode, final boolean truncateLastFile) throws IOException {
+        beforeEach();
 
-            final Random random = RandomUtils.getRandomPrintSeed();
+        final Random random = RandomUtils.getRandomPrintSeed();
 
-            final PlatformContext platformContext = buildContext(ancientMode);
+        final PlatformContext platformContext = buildContext(ancientMode);
 
-            final StandardGraphGenerator generator = PcesWriterTestUtils.buildGraphGenerator(platformContext, random);
-            final int stepsUntilAncient = random.nextInt(50, 100);
-            final PcesSequencer sequencer = new DefaultPcesSequencer();
-            final PcesFileTracker pcesFiles = new PcesFileTracker(ancientMode);
+        final StandardGraphGenerator generator = PcesWriterTestUtils.buildGraphGenerator(platformContext, random);
+        final int stepsUntilAncient = random.nextInt(50, 100);
+        final PcesSequencer sequencer = new DefaultPcesSequencer();
+        final PcesFileTracker pcesFiles = new PcesFileTracker(ancientMode);
 
-            final PcesFileManager fileManager = new PcesFileManager(platformContext, pcesFiles, selfId, 0);
-            final DefaultPcesWriter writer = new DefaultPcesWriter(platformContext, fileManager);
-            final AtomicLong latestDurableSequenceNumber = new AtomicLong();
+        final PcesFileManager fileManager = new PcesFileManager(platformContext, pcesFiles, selfId, 0);
+        final DefaultPcesWriter writer = new DefaultPcesWriter(platformContext, fileManager);
+        final AtomicLong latestDurableSequenceNumber = new AtomicLong();
 
-            final List<PlatformEvent> eventsBeforeDiscontinuity = new LinkedList<>();
-            final List<PlatformEvent> eventsAfterDiscontinuity = new LinkedList<>();
-            for (int i = 0; i < numEvents; i++) {
-                final PlatformEvent event =
-                        generator.generateEventWithoutIndex().getBaseEvent();
-                if (i < numEvents / 2) {
-                    eventsBeforeDiscontinuity.add(event);
-                } else {
-                    eventsAfterDiscontinuity.add(event);
-                }
+        final List<PlatformEvent> eventsBeforeDiscontinuity = new LinkedList<>();
+        final List<PlatformEvent> eventsAfterDiscontinuity = new LinkedList<>();
+        for (int i = 0; i < numEvents; i++) {
+            final PlatformEvent event =
+                    generator.generateEventWithoutIndex().getBaseEvent();
+            if (i < numEvents / 2) {
+                eventsBeforeDiscontinuity.add(event);
+            } else {
+                eventsAfterDiscontinuity.add(event);
             }
-
-            writer.beginStreamingNewEvents();
-
-            final Collection<PlatformEvent> rejectedEvents = new HashSet<>();
-
-            long lowerBound = ancientMode.selectIndicator(0, 1);
-            final Iterator<PlatformEvent> iterator1 = eventsBeforeDiscontinuity.iterator();
-            while (iterator1.hasNext()) {
-                final PlatformEvent event = iterator1.next();
-
-                sequencer.assignStreamSequenceNumber(event);
-                passValueToDurabilityNexus(writer.writeEvent(event), latestDurableSequenceNumber);
-
-                lowerBound = Math.max(lowerBound, event.getAncientIndicator(ancientMode) - stepsUntilAncient);
-                writer.updateNonAncientEventBoundary(new EventWindow(1, lowerBound, lowerBound, ancientMode));
-
-                if (event.getAncientIndicator(ancientMode) < lowerBound) {
-                    // Although it's not common, it's actually possible that the generator will generate
-                    // an event that is ancient (since it isn't aware of what we consider to be ancient)
-                    rejectedEvents.add(event);
-                    iterator1.remove();
-                }
-
-                // request a flush sometimes
-                if (random.nextInt(10) == 0) {
-                    passValueToDurabilityNexus(
-                            writer.submitFlushRequest(event.getStreamSequenceNumber()), latestDurableSequenceNumber);
-                }
-            }
-
-            passValueToDurabilityNexus(writer.registerDiscontinuity(100L), latestDurableSequenceNumber);
-            eventsBeforeDiscontinuity.forEach(
-                    event -> assertTrue(latestDurableSequenceNumber.get() >= event.getStreamSequenceNumber()));
-
-            if (truncateLastFile) {
-                // Remove a single byte from the last file. This will corrupt the last event that was written.
-                final Iterator<PcesFile> it = pcesFiles.getFileIterator(NO_LOWER_BOUND, 0);
-                while (it.hasNext()) {
-                    final PcesFile file = it.next();
-                    if (!it.hasNext()) {
-                        FileManipulation.truncateNBytesFromFile(file.getPath(), 1);
-                    }
-                }
-
-                eventsBeforeDiscontinuity.remove(eventsBeforeDiscontinuity.size() - 1);
-            }
-
-            final Iterator<PlatformEvent> iterator2 = eventsAfterDiscontinuity.iterator();
-            while (iterator2.hasNext()) {
-                final PlatformEvent event = iterator2.next();
-
-                sequencer.assignStreamSequenceNumber(event);
-                passValueToDurabilityNexus(writer.writeEvent(event), latestDurableSequenceNumber);
-
-                lowerBound = Math.max(lowerBound, event.getAncientIndicator(ancientMode) - stepsUntilAncient);
-                writer.updateNonAncientEventBoundary(new EventWindow(1, lowerBound, lowerBound, ancientMode));
-
-                if (event.getAncientIndicator(ancientMode) < lowerBound) {
-                    // Although it's not common, it's actually possible that the generator will generate
-                    // an event that is ancient (since it isn't aware of what we consider to be ancient)
-                    rejectedEvents.add(event);
-                    iterator2.remove();
-                }
-
-                // request a flush sometimes
-                if (random.nextInt(10) == 0) {
-                    passValueToDurabilityNexus(
-                            writer.submitFlushRequest(event.getStreamSequenceNumber()), latestDurableSequenceNumber);
-                }
-            }
-
-            passValueToDurabilityNexus(
-                    writer.submitFlushRequest(eventsAfterDiscontinuity.getLast().getStreamSequenceNumber()),
-                    latestDurableSequenceNumber);
-
-            assertTrue(latestDurableSequenceNumber.get()
-                    >= eventsAfterDiscontinuity
-                            .get(eventsAfterDiscontinuity.size() - 1)
-                            .getStreamSequenceNumber());
-            eventsAfterDiscontinuity.forEach(
-                    event -> assertTrue(latestDurableSequenceNumber.get() >= event.getStreamSequenceNumber()));
-            rejectedEvents.forEach(
-                    event -> assertFalse(latestDurableSequenceNumber.get() >= event.getStreamSequenceNumber()));
-
-            PcesWriterTestUtils.verifyStream(
-                    selfId, eventsBeforeDiscontinuity, platformContext, truncateLastFile ? 1 : 0, ancientMode);
-
-            writer.closeCurrentMutableFile();
         }
+
+        writer.beginStreamingNewEvents();
+
+        final Collection<PlatformEvent> rejectedEvents = new HashSet<>();
+
+        long lowerBound = ancientMode.selectIndicator(0, 1);
+        final Iterator<PlatformEvent> iterator1 = eventsBeforeDiscontinuity.iterator();
+        while (iterator1.hasNext()) {
+            final PlatformEvent event = iterator1.next();
+
+            sequencer.assignStreamSequenceNumber(event);
+            passValueToDurabilityNexus(writer.writeEvent(event), latestDurableSequenceNumber);
+
+            lowerBound = Math.max(lowerBound, event.getAncientIndicator(ancientMode) - stepsUntilAncient);
+            writer.updateNonAncientEventBoundary(new EventWindow(1, lowerBound, lowerBound, ancientMode));
+
+            if (event.getAncientIndicator(ancientMode) < lowerBound) {
+                // Although it's not common, it's actually possible that the generator will generate
+                // an event that is ancient (since it isn't aware of what we consider to be ancient)
+                rejectedEvents.add(event);
+                iterator1.remove();
+            }
+
+            // request a flush sometimes
+            if (random.nextInt(10) == 0) {
+                passValueToDurabilityNexus(
+                        writer.submitFlushRequest(event.getStreamSequenceNumber()), latestDurableSequenceNumber);
+            }
+        }
+
+        passValueToDurabilityNexus(writer.registerDiscontinuity(100L), latestDurableSequenceNumber);
+        eventsBeforeDiscontinuity.forEach(
+                event -> assertTrue(latestDurableSequenceNumber.get() >= event.getStreamSequenceNumber()));
+
+        if (truncateLastFile) {
+            // Remove a single byte from the last file. This will corrupt the last event that was written.
+            final Iterator<PcesFile> it = pcesFiles.getFileIterator(NO_LOWER_BOUND, 0);
+            while (it.hasNext()) {
+                final PcesFile file = it.next();
+                if (!it.hasNext()) {
+                    FileManipulation.truncateNBytesFromFile(file.getPath(), 1);
+                }
+            }
+
+            eventsBeforeDiscontinuity.remove(eventsBeforeDiscontinuity.size() - 1);
+        }
+
+        final Iterator<PlatformEvent> iterator2 = eventsAfterDiscontinuity.iterator();
+        while (iterator2.hasNext()) {
+            final PlatformEvent event = iterator2.next();
+
+            sequencer.assignStreamSequenceNumber(event);
+            passValueToDurabilityNexus(writer.writeEvent(event), latestDurableSequenceNumber);
+
+            lowerBound = Math.max(lowerBound, event.getAncientIndicator(ancientMode) - stepsUntilAncient);
+            writer.updateNonAncientEventBoundary(new EventWindow(1, lowerBound, lowerBound, ancientMode));
+
+            if (event.getAncientIndicator(ancientMode) < lowerBound) {
+                // Although it's not common, it's actually possible that the generator will generate
+                // an event that is ancient (since it isn't aware of what we consider to be ancient)
+                rejectedEvents.add(event);
+                iterator2.remove();
+            }
+
+            // request a flush sometimes
+            if (random.nextInt(10) == 0) {
+                passValueToDurabilityNexus(
+                        writer.submitFlushRequest(event.getStreamSequenceNumber()), latestDurableSequenceNumber);
+            }
+        }
+
+        passValueToDurabilityNexus(
+                writer.submitFlushRequest(eventsAfterDiscontinuity.getLast().getStreamSequenceNumber()),
+                latestDurableSequenceNumber);
+
+        assertTrue(latestDurableSequenceNumber.get()
+                >= eventsAfterDiscontinuity
+                .get(eventsAfterDiscontinuity.size() - 1)
+                .getStreamSequenceNumber());
+        eventsAfterDiscontinuity.forEach(
+                event -> assertTrue(latestDurableSequenceNumber.get() >= event.getStreamSequenceNumber()));
+        rejectedEvents.forEach(
+                event -> assertFalse(latestDurableSequenceNumber.get() >= event.getStreamSequenceNumber()));
+
+        PcesWriterTestUtils.verifyStream(
+                selfId, eventsBeforeDiscontinuity, platformContext, truncateLastFile ? 1 : 0, ancientMode);
+
+        writer.closeCurrentMutableFile();
+
     }
 
     /**

--- a/platform-sdk/swirlds-unit-tests/core/swirlds-platform-test/src/test/java/com/swirlds/platform/test/event/preconsensus/PcesWriterTests.java
+++ b/platform-sdk/swirlds-unit-tests/core/swirlds-platform-test/src/test/java/com/swirlds/platform/test/event/preconsensus/PcesWriterTests.java
@@ -1,5 +1,5 @@
 /*
- * Copyright (C) 2023-2024 Hedera Hashgraph, LLC
+ * Copyright (C) 2023-2025 Hedera Hashgraph, LLC
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
@@ -401,10 +401,10 @@ class PcesWriterTests {
 
     @ParameterizedTest
     @CsvSource({
-            "GENERATION_THRESHOLD, true",
-            "GENERATION_THRESHOLD, false",
-            "BIRTH_ROUND_THRESHOLD, true",
-            "BIRTH_ROUND_THRESHOLD, false"
+        "GENERATION_THRESHOLD, true",
+        "GENERATION_THRESHOLD, false",
+        "BIRTH_ROUND_THRESHOLD, true",
+        "BIRTH_ROUND_THRESHOLD, false"
     })
     @DisplayName("Discontinuity Test")
     void discontinuityTest(@NonNull final AncientMode ancientMode, final boolean truncateLastFile) throws IOException {
@@ -426,8 +426,7 @@ class PcesWriterTests {
         final List<PlatformEvent> eventsBeforeDiscontinuity = new LinkedList<>();
         final List<PlatformEvent> eventsAfterDiscontinuity = new LinkedList<>();
         for (int i = 0; i < numEvents; i++) {
-            final PlatformEvent event =
-                    generator.generateEventWithoutIndex().getBaseEvent();
+            final PlatformEvent event = generator.generateEventWithoutIndex().getBaseEvent();
             if (i < numEvents / 2) {
                 eventsBeforeDiscontinuity.add(event);
             } else {
@@ -511,8 +510,8 @@ class PcesWriterTests {
 
         assertTrue(latestDurableSequenceNumber.get()
                 >= eventsAfterDiscontinuity
-                .get(eventsAfterDiscontinuity.size() - 1)
-                .getStreamSequenceNumber());
+                        .get(eventsAfterDiscontinuity.size() - 1)
+                        .getStreamSequenceNumber());
         eventsAfterDiscontinuity.forEach(
                 event -> assertTrue(latestDurableSequenceNumber.get() >= event.getStreamSequenceNumber()));
         rejectedEvents.forEach(
@@ -522,7 +521,6 @@ class PcesWriterTests {
                 selfId, eventsBeforeDiscontinuity, platformContext, truncateLastFile ? 1 : 0, ancientMode);
 
         writer.closeCurrentMutableFile();
-
     }
 
     /**

